Implement CGDiff

silas-martens opened the pull request:
This MR introduces CGDiff, a utility to compare two MetaCG call graphs.
The comparison is based on features such as name, hasBody, callees, and metadata.

**NOTE**: Since nodeIDs are not unique across call graphs, the node name is used as the unique identifier.

*Features:*
- Compare 2 callgraphs based on name, `hasBody`, `callees`, `metadata`
- metadata and callees are compared order-independant using string sets
- selectively ignore specific comparison features:
  - `--ignore-body`: ignore body
  - `--ignore-edges`: ignore edges
  - `--ignore-md`: ignore metadata
  - `--ignore-edge-md`: ignore edge metadata
  - `--ignore-global-md`: ignore global metadata
  - `--ignore-md-key arg`: ignore specific metadata
  - `--emit-diff-on-success`: emit diff on success (by default only emitted when call-graphs differ)
- Return diff file as json

I aimed for efficiency, but the tool has not yet been tested on large, real-world applications. Feedback and suggestions are very welcome!

# Output example

```
{
    "cgA": "/home/silasm/work/github/MetaCG/build2/../tools/cgdiff/test/integration/input/v4_global_md.mcg",
    "cgB": "/home/silasm/work/github/MetaCG/build2/../tools/cgdiff/test/integration/input/v4_global_md_1.mcg",
    "ignoring": [
        "metadata"
    ],
    "diff": {
        "globalMetaDiff": {
            "entryFunction": {
                "aValue": "\"thisIsMain\"",
                "bValue": "\"thisIsNotMain\""
            }
        },
        "nodeDiffs": {
            "thisIsNotMain": {
                "diffType": [
                    "differentEdges"
                ],
                "in": "both",
                "metadataOnlyInA": [],
                "metadataOnlyInB": [],
                "edges": {
                    "foo": {
                        "in": "cgB",
                        "metadataOnlyInA": [],
                        "metadataOnlyInB": []
                    }
                }
            },
            "foo": {
                "diffType": [
                    "missingNode"
                ],
                "in": "cgA",
                "metadataOnlyInA": [],
                "metadataOnlyInB": []
            }
        }
    }
}
```

silas-martens committed:
Implement CGDiff

silas-martens committed:
Added output to file

silas-martens committed:
sort json output

jplehr commented on tools/cgdiff/include/NodeSummaryHasher.h:
```diff
@@ -0,0 +1,50 @@
+ * The hash is order-independent for sets (callees, metadata) using XOR,
+ * and combines individual field hashes using boost's hash-combine.
+ */
+struct NodeSummaryHasher {
```
```suggestion
class NodeSummaryHasher {
```
If we use `public` and `private`, I think we should use `class`.

jplehr commented on tools/cgdiff/include/NodeSummaryHasher.h:
```diff
@@ -0,0 +1,50 @@
+            hash_combine(seed, ns.hasBody);
+        }
+        if (!hasFlag(mode, ignoreEdges)) {
+            // callees ist unordered_set<string>
```
```suggestion
            // callees is unordered_set<string>
```

jplehr commented on tools/cgdiff/include/NodeSummary.h:
```diff
@@ -0,0 +1,206 @@
+    NodeDiff diff;
+    diff.name = nsA.name;
+
+    if (!hasFlag(mode, ignoreBody) && nsA.hasBody != nsB.hasBody) {
```
I think that adding parens is helpful here and further down.
```suggestion
    if (!hasFlag(mode, ignoreBody) && (nsA.hasBody != nsB.hasBody)) {
```

jplehr commented on tools/cgdiff/include/NodeSummary.h:
```diff
@@ -0,0 +1,206 @@
+            }
+        }
+        for (const auto& callee : nsB.callees)
+            if (!nsA.callees.count(callee))
+                diff.calleesOnlyInB.insert(callee);
```
```suggestion
            if (!nsA.callees.count(callee)) {
                diff.calleesOnlyInB.insert(callee);
            }
```
I think we always use curlys

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,227 @@
+ * https://github.com/tudasc/metacg/LICENSE.txt
+ */
+
+#include "io/MCGReader.h"
+#include <cxxopts.hpp>
+#include "CgNode.h"
+#include "NodeSummary.h"
+#include "NodeSummaryComparator.h"
+#include "NodeSummaryHasher.h"
```
First local includes, then project, then deps, then STL.

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,227 @@
+            std::cout << options.help() << std::endl;
+            exit(0);
+        }
+
```
```suggestion
```

jplehr reviewed (commented):
Some initial comments

sebastiankreutzer commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,227 @@
+
+    // Compare nodes
+    // A
+    Set nodesA(0, nameHasher, nameComparator);
```
This code is duplicated for A and B. Please refactor by implementing this in a separate function or lambda.

sebastiankreutzer commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,227 @@
+                "Returns 0 if call graphs are equal, 1 otherwise.\n");
+
+        options.add_options()
+            ("ignoreEdges", "ignore Edges", cxxopts::value<bool>()->default_value("false"))
```
This is not a big issue, but personally I prefer the `option-name` format to camel case for program options.

sebastiankreutzer reviewed (commented):
Thanks for the contribution!

pearzt commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,227 @@
+#include "NodeSummaryHasher.h"
+
+namespace metacg { 
+    struct NodeNameToStrMapping : NodeToStrMapping {
```
There is some code duplication here with pymetacg's `NameMapping`. Could you please extract this to a common header?

silas-martens committed:
extract duplicated code into lambda

silas-martens committed:
renamed commandline options

silas-martens committed:
reordered includes

silas-martens committed:
changed struct -> class

silas-martens committed:
moved NameMapping into its own header

silas-martens committed:
updated integration tests with renamed options

silas-martens committed:
add callgraph paths to diff output

silas-martens committed:
update usages of namemapping

silas-martens committed:
Apply suggestions from code review

Co-authored-by: Jan Patrick Lehr <jp.lehr@gmail.com>

jplehr commented on tools/cgdiff/test/integration/input/whole.mcg:
Does this giant file add any value? Should probably be removed.

jplehr commented on graph/include/io/NameMapping.h:
```diff
@@ -0,0 +1,24 @@
+namespace metacg {
+    struct NameMapping : metacg::NodeToStrMapping {
+        public:
+            NameMapping(const metacg::Callgraph& graph) : _graph(graph) {}
```
Should this be marked `explicit`?

jplehr commented on graph/include/io/NameMapping.h:
```diff
@@ -0,0 +1,24 @@
+            virtual std::string getStrFromNode(metacg::NodeId id) override { return _graph.getNode(id)->getFunctionName(); }
+
+        private:
+            const metacg::Callgraph& _graph;
```
```suggestion
            const metacg::Callgraph& graph;
```

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+
+std::vector<NodeDiff> compare(const metacg::Callgraph& mcgA,
+        const metacg::Callgraph& mcgB, ComparisonMode mode) {
+    using Set = std::unordered_set<NodeSummary, NodeSummaryHasher, NodeSummaryComparator>;
```
Can this be named more expressively?

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+                            std::move(calleeNames)));
+            }
+        }
+
```
```suggestion
```

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+
+            // Metadata
+            if (!hasFlag(mode, ignoreMetadata)) {
+                std::unordered_set<std::string> metadataList;
```
Why call it list? Isn't this just `metadata`?

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+            } else {
+                nodes.insert(NodeSummary(node->getFunctionName(), node->getHasBody(), 
+                            std::move(calleeNames)));
+            }
```
Isn't this the same as the `if (true)` case with the only difference that `metdata.empty() == true`?
This can be simplified and no `else` is required.

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+        else {
+            nodesB.erase(itB); 
+        }
+
```
```suggestion
```

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+            diffs.emplace_back(createNodeDiff(nA, *itB, mode)); 
+            nodesB.erase(itB); 
+        }
+        else {
```
```suggestion
        } else {
```

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+
+        if (result.count("help")) {
+            std::cout << options.help() << std::endl;
+            exit(0);
```
Should this be `return 0;` instead?

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,192 @@
+        return 1;
+    } catch (const std::exception& e) {
+        std::cerr << "Unexpected error: " << e.what() << std::endl;
+        return 1;
```
Consider using other return codes in exceptional case.

jplehr reviewed (commented):
Another couple of drive-by comments.

silas-martens committed:
remove whole mcg

silas-martens committed:
first version of edge metadata comparison

silas-martens committed:
Add initial support for edge metadata and global metadata.
- small code refactor
- updated diff format
- text output is wip and not available for now.

silas-martens committed:
fixed bug when comparing metadata values

silas-martens committed:
reworked integration tests

jplehr commented on tools/cgdiff/test/unit/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+
+add_metacg(main)
+add_googletest_libraries(main)
```
Please make the unit test discoverable with `ctest`

silas-martens committed:
bugfix & clean up

silas-martens committed:
added option to ignore specific metadata

silas-martens committed:
Fixed bug

silas-martens committed:
removed text output and reworked options

silas-martens committed:
made unit tests discoverable with ctest

silas-martens committed:
updated integration tests

silas-martens committed:
Made constructor explicit

silas-martens committed:
Small fixes

silas-martens committed:
Fixe ci

silas-martens committed:
Added groundtruthes

silas-martens committed:
fixed typo

silas-martens committed:
removed std::cout

silas-martens committed:
Integrated gitlab ci

silas-martens committed:
sorted includes

silas-martens committed:
Added integration test README.md

silas-martens committed:
Add README

silas-martens committed:
Write test output into build dir to avoid race conditions

silas-martens committed:
Added text output indicating wether files differ or not

silas-martens committed:
Reformat CMakeLists.txt

silas-martens committed:
Reformat files

silas-martens commented:
I added global and edge metadata, and reworked the diff file format and options. Your previous feedback should now be addressed. Feel free to share your thoughts.

silas-martens committed:
Revert changes to cgpatch as they will be included in a later pr

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,262 @@
+          std::unordered_set<std::string> emd;
+          for (auto& metadata : mcg.getAllEdgeMetaData(*node, *callee)) {
+            std::string key = metadata.first;
+            std::string value = metadata.second->toJson(mapping).dump(-1);
```
What's this magic number?

silas-martens replied:
The argument of dump specifies the indentation, -1 selects the most compact representation. As this is the default i will remove this.

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,262 @@
+                             "Returns 0 if call-graphs are equal, 1 otherwise.\n");
+
+    options.add_options()("ignore-edges", "ignore edges", cxxopts::value<bool>()->default_value("false"))(
+        "ignore-body", "ignore body", cxxopts::value<bool>()->default_value("false"))(
```
What does the body refer to here? Does this refer to the `hasBody` data?

silas-martens replied:
Yes this refers to `hasBody`, I will change the option to `ignore-hasBody` to make that clearer.

jplehr commented on tools/cgdiff/CGDiff.cpp:
```diff
@@ -0,0 +1,262 @@
+    if (argc < 3) {
+      std::cerr << "Usage: ./cgdiff [options] <cg1.json> <cg2.json>\n";
+      return 1;
+    }
```
Should this go before the `cxxopts` stuff above?

silas-martens replied:
Oh, this is actually redundant with the following if statement. This was a workaround, I used before we updated cxxopts.

jplehr commented on tools/cgdiff/include/Diff.h:
```diff
@@ -0,0 +1,14 @@
+ * License: Part of the MetaCG project. Licensed under BSD 3 clause license. See LICENSE.txt file at
+ * https://github.com/tudasc/metacg/LICENSE.txt
+ */
+#pragma once
```
Should we maintain using include guards or rely on `pragma once`?

silas-martens replied:
I’m fine with either. I used #pragma once mostly for convenience.

jplehr commented on tools/cgdiff/include/Diff.h:
```diff
@@ -0,0 +1,14 @@
+#pragma once
+#include "nlohmann/json.hpp"
+
+struct Diff {
```
Should this be inside a/the `metacg` namespace?

jplehr commented on tools/cgdiff/include/DiffFormatter.h:
```diff
@@ -0,0 +1,155 @@
+/**
+ * File: CGDiff.cpp
```
```suggestion
 * File: DiffFormatter.h
```

jplehr commented on tools/cgdiff/include/DiffFormatter.h:
```diff
@@ -0,0 +1,155 @@
+    nlohmann::ordered_json globalMetaDiffs = nlohmann::ordered_json::object();
+
+    for (const auto& diff : diffs) {
+      nlohmann::ordered_json inner = diff->toJson();  // each Diff subclass implements its own toJson()
```
```suggestion
      // each Diff subclass implements its own toJson()
      nlohmann::ordered_json inner = diff->toJson();
```
I personally prefer this way to attach comments. Gives the formatter a better time.

jplehr commented on tools/cgdiff/include/DiffFormatter.h:
```diff
@@ -0,0 +1,155 @@
+          globalMetaDiffs[md->key] = inner;
+          break;
+        }
+      }
```
Should this have a `default` with some `assert` or `__builtin_unreachable` to guard against future mistakes?

jplehr commented on tools/cgdiff/include/DiffFormatter.h:
```diff
@@ -0,0 +1,155 @@
+ * @return NodeDiff describing the differences between nsA and nsB
+ */
+NodeDiff createNodeDiff(const NodeSummary& nsA, const NodeSummary& nsB, ComparisonMode mode) {
+  NodeDiff diff;
```
Do we want to allow "unnamed diffs"? If not, can we make this a CTor argument?

jplehr commented on tools/cgdiff/include/GlobalMDDiff.h:
```diff
@@ -0,0 +1,26 @@
+/**
+ * File: DiffFormatter.cpp
```
```suggestion
 * File: GlobalMDDiff.h
```

jplehr commented on tools/cgdiff/include/NodeDiff.h:
```diff
@@ -0,0 +1,89 @@
+/**
+ * File: DiffFormatter.cpp
```
```suggestion
 * File: NodeDiff.h
```

jplehr commented on tools/cgdiff/include/NodeDiff.h:
```diff
@@ -0,0 +1,89 @@
+  };
+
+  NodeDiff() = default;
+  NodeDiff(std::string name_, std::vector<std::string> diffType_ = {}, std::vector<EdgeDiff> edgeDiffs_ = {},
```
```suggestion
  NodeDiff(std::string name, std::vector<std::string> diffType = {}, std::vector<EdgeDiff> edgeDiffs = {},
```

jplehr commented on tools/cgdiff/include/NodeDiff.h:
```diff
@@ -0,0 +1,89 @@
+
+  NodeDiff() = default;
+  NodeDiff(std::string name_, std::vector<std::string> diffType_ = {}, std::vector<EdgeDiff> edgeDiffs_ = {},
+           std::unordered_set<std::string> metaA_ = {}, std::unordered_set<std::string> metaB_ = {})
```
```suggestion
           std::unordered_set<std::string> metaA = {}, std::unordered_set<std::string> metaB = {})
```

jplehr commented on tools/cgdiff/include/NodeDiff.h:
```diff
@@ -0,0 +1,89 @@
+   * @param a NodeSummary from the first graph
+   * @return NodeDiff representing a missing node in A
+   */
+  static NodeDiff onlyInA(const NodeSummary& a) {
```
Does this do anything with the `NodeSummary` other than reading the name?

silas-martens replied:
No, it only reads the name.

jplehr commented on tools/cgdiff/include/NodeSummary.h:
```diff
@@ -0,0 +1,63 @@
+struct NodeSummary {
+  std::string name;
+  bool hasBody;
+  std::unordered_set<std::string> callees;  // use function names as unique identifier
```
```suggestion
  // use function names as unique identifier
  std::unordered_set<std::string> callees;
```
Please here and elsewhere.

jplehr commented on tools/cgdiff/test/integration/groundtruthes/body_different.json:
```diff
@@ -0,0 +1,18 @@
+{
+    "cgA": "/home/silasm/work/github/MetaCG/build2/../tools/cgdiff/test/integration/input/cgA_basic.mcg",
+    "cgB": "/home/silasm/work/github/MetaCG/build2/../tools/cgdiff/test/integration/input/cgD_basic.mcg",
```
Should we again pay some attention to the path names in here?

silas-martens committed:
Fixed typo

Co-authored-by: Jan Patrick Lehr <jp.lehr@gmail.com>

silas-martens committed:
Reformat comment

Co-authored-by: Jan Patrick Lehr <jp.lehr@gmail.com>

silas-martens committed:
Rename set

Co-authored-by: Jan Patrick Lehr <jp.lehr@gmail.com>

silas-martens committed:
Rename arg

Co-authored-by: Jan Patrick Lehr <jp.lehr@gmail.com>

silas-martens committed:
Apply suggestions from code review

Co-authored-by: Jan Patrick Lehr <jp.lehr@gmail.com>

silas-martens committed:
Renamed ignore-Body option to ignore-hasBody

silas-martens committed:
Added assert to guard against future mistakes

silas-martens committed:
Removed default ctor

silas-martens committed:
Wrapped Diff in metacg namespace

jplehr commented on tools/cgdiff/test/integration/input/README.md:
Pretty sure this will be unmaintained rather quickly.

pearzt replied:
Probably true, but it's still nice to have this overview somewhere. I think my vote is for keeping this and trying to remember that it needs to be updated when somebody touches the CGDiff tests.

jplehr replied:
Yeah, I wasn't voting to delete it, but wanted to point out that such things end up unmaintained quickly

jplehr reviewed (approved):
LGTM

One other remark and not a blocker for this to land: Should machinery (e.g., classes / functions) that is local to a MetaCG tool (e.g., CGDiff) be also inside the MetaCG namespace?
